stl/inc/span

```diff
@@ -226,28 +226,28 @@ struct pointer_traits<_Span_iterator<_Ty>> {
 };
 
 // STRUCT TEMPLATE _Span_extent_type
-template <size_t _Extent>
+template <class _Ty, size_t _Extent>
 struct _Span_extent_type {
+    using pointer = _Ty*;
+
     constexpr _Span_extent_type() noexcept = default;
 
-    constexpr explicit _Span_extent_type(size_t) noexcept {}
+    constexpr explicit _Span_extent_type(size_t, const pointer _Data) noexcept : _Mydata{_Data} {}
 
-    _NODISCARD constexpr size_t size() const noexcept {
-        return _Extent;
-    }
+    pointer _Mydata{nullptr};
+    static constexpr size_t _Mysize = _Extent;
 };
 
-template <>
-struct _Span_extent_type<dynamic_extent> {
-    constexpr _Span_extent_type() noexcept = default;
+template <class _Ty>
+struct _Span_extent_type<_Ty, dynamic_extent> {
+    using pointer = _Ty*;
 
-    constexpr explicit _Span_extent_type(const size_t _Size) noexcept : _Mysize(_Size) {}
+    constexpr _Span_extent_type() noexcept = default;
 
-    _NODISCARD constexpr size_t size() const noexcept {
-        return _Mysize;
-    }
+    constexpr explicit _Span_extent_type(const size_t _Size, const pointer _Data) noexcept
+        : _Mydata{_Data}, _Mysize{_Size} {}
 
-private:
+    pointer _Mydata{nullptr};

```

Comment:
I want to note that [libstdc++](https://github.com/gcc-mirror/gcc/blob/53d461e4e0fbc6d2ef7385509e8c3785e9b205fd/libstdc%2B%2B-v3/include/std/span#L398-L399) does it the other way around (size + ptr) whereas [libc++](https://github.com/llvm/llvm-project/blob/1778564f9118e29628a76a9cfea87d76fa83f0e6/libcxx/include/span#L512-L513) uses the same convention (ptr - size)

Reply:
Thanks - that appears to substantially invalidate the original motivation in #874. I still think this PR is worth accepting (although perhaps the test comment that we want span to be standard_layout "for portability" is no longer justified).

Reply:
I have asked Jonathan to undo his layout decision for libstdc++ at https://gcc.gnu.org/bugzilla/show_bug.cgi?id=95609. I expect that it can no longer be honoured, due to libstdc++'s strong backwards ABI commitment. But it is worth asking anyway.

Re the bigger picture, the Windows kernel may or may not extend Windows RIO into a generalised io_uring-like kernel API for all i/o. If they do, the native `struct iovec` for Windows would actually be a `{size_t, size_t}` into a registered buffer. If `std::file_handle` is standardised, we would map that by implementing within the standard library buffer copying using `memcpy` into the registered buffers. So, if that comes to pass, a ptr + size layout wouldn't be useful.

However, the Windows kernel may instead implement a more generalised io_uring-like kernel API for all i/o, and if so, I shall be asking for ptr + size layout so the C++ scatter-gather buffers can be zero copied to the kernel. I appreciate that this may seem over eager optimisation, however if your i/o quantum is a few thousand bytes made up of a few hundred scatter-gather buffers, eliminating the buffer translation copy makes an *enormous* difference to overall performance. A niche use case, sure, but for some users a very important one.